ref_arch/multi_tenant.rst

```diff
@@ -0,0 +1,576 @@
+
+::
+
+  SELECT latlon
+    FROM geo_ips, clicks c
+   WHERE addrs >> c.user_ip
+     AND c.clicked_at > current_date - INTERVAL '1 day'
+     AND c.company_id = 5
+     AND c.ad_id = 456;
+
+To run this query efficiently in a distributed setup, we need to find a way to co-locate the :code:`geo_ips` table with clicks for every company. That way, no network traffic need be incurred at query time. One way would be to add a :code:`company_id` column to :code:`geo_ips` and duplicate the data in the table for every company. This approach is not optimal because it introduces the burden of keeping the data synchronized between the companies if and when it changes. A more convenient way is by designating :code:`geo_ips` as a *reference table.*
+
+Reference tables are replicated across all worker nodes, and Citus automatically keeps them in sync during modifications. This co-locates the information for all tenants' queries. To make a reference table, call :ref:`create_reference_table <create_reference_table>` for a table on the coordinator node:
+
+::
+
+  -- Make synchronized copies of geo_ips on all workers
+
+  SELECT create_reference_table('geo_ips');
+
+After doing this, the join query (presented earlier) on :code:`geo_ips` and :code:`clicks` will perform efficiently.
+
+Online Changes to the Schema
+----------------------------
+
+Another challenge with multi-tenant systems is keeping the schemas for all the tenants in sync. Any schema change needs to be consistently reflected across all the tenants. In Citus, you can use standard postgres DDL commands to change the schema of your tables, and Citus will propagate them from the coordinator node to the workers using a two-phase commit protocol.
+
+For example, the advertisements in this application could use a text caption. We can add a column to the table by issuing the standard SQL on the coordinator:
+
+::
+
+  ALTER TABLE ads
+    ADD COLUMN caption text;
+
+This updates all the workers as well. Once this command finishes, the Citus cluster will accept queries that read or write data in the new :code:`caption` column.
+
+For a fuller explanation of which DDL commands propagate through the cluster, see :ref:`ddl_prop_support`.
+
+Per-Tenant Customization
+========================
+
+Given that all tenants share a common schema and hardware infrastructure, how can we accommodate tenants which want to store information not needed by others? For example, one of the tenant applications using our advertising database may want to store tracking cookie information with clicks, whereas another tenant may care about browser agents. Traditionally databases using a shared schema approach for multi-tenancy have resorted to creating a fixed number of pre-allocated "custom" columns, or having external "extension tables." However PostgreSQL provides a much easier way with its unstructured column types, notably `JSONB <https://www.postgresql.org/docs/current/static/datatype-json.html>`_.
```

Comment:
Are you sure? I'm more familiar with the period going inside the quotes. The [APA style guide](http://blog.apastyle.org/apastyle/2011/08/punctuating-around-quotation-marks.html) recommends that style for general usage as well.